internal/storage/memory.go

```diff
@@ -361,3 +366,175 @@ func inMemoryCRDB() (testserver.TestServer, error) {
+		db:         config.db,
+		httpClient: http.DefaultClient,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	err := s.createTables()
+
+	return s, err
+}
+
+// WithHTTPClient allows configuring the HTTP client used by
+// memoryUserInfoService to call out to userinfo endpoints.
+func WithHTTPClient(client *http.Client) func(svc *memoryUserInfoService) {
+	return func(svc *memoryUserInfoService) {
+		svc.httpClient = client
+	}
+}
+
+func (s *memoryUserInfoService) createTables() error {
+	_, err := s.db.Exec(`
+        CREATE TABLE IF NOT EXISTS user_info (
+            id    UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+            name  STRING,
+            email STRING,
+            sub   STRING NOT NULL,
+            iss_id   UUID NOT NULL REFERENCES issuers(id),
+            UNIQUE (iss_id, sub)
+        )`)
+
+	return err
+}
+
+// LookupUserInfoByClaims fetches UserInfo from the store.
+// This does not make an HTTP call with the subject token, so for this
+// data to be available, the data must have already be fetched and
+// stored.
+func (s memoryUserInfoService) LookupUserInfoByClaims(ctx context.Context, iss, sub string) (*types.UserInfo, error) {
+	row := s.db.QueryRowContext(ctx, `
+        SELECT ui.name, ui.email, ui.sub, i.uri FROM user_info ui
+        JOIN issuers i ON
+           ui.iss_id = i.id
+        WHERE
+           i.uri = $1 AND ui.sub = $2
+        `, iss, sub)
+
+	var ui types.UserInfo
+
+	err := row.Scan(&ui.Name, &ui.Email, &ui.Subject, &ui.Issuer)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, types.ErrUserInfoNotFound
+	}
+
+	return &ui, err
+}
+
+func (s memoryUserInfoService) LookupUserInfoByID(ctx context.Context, id string) (*types.UserInfo, error) {
+	row := s.db.QueryRowContext(ctx, `
+        SELECT ui.id, ui.name, ui.email, ui.sub, i.uri FROM user_info ui
+        JOIN issuers i ON
+           ui.iss_id = i.id
+        WHERE
+           ui.id = $1
+        `, id)
+
+	var ui types.UserInfo
+
+	err := row.Scan(&ui.ID, &ui.Name, &ui.Email, &ui.Subject, &ui.Issuer)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, types.ErrUserInfoNotFound
+	}
+
+	return &ui, err
+}
+
+// StoreUserInfo is used to store user information by issuer and
+// subject pairs. UserInfo is unique to issuer/subject pairs.
+func (s memoryUserInfoService) StoreUserInfo(ctx context.Context, userInfo types.UserInfo) (*types.UserInfo, error) {
+	row := s.db.QueryRowContext(ctx, `
+        SELECT id FROM issuers WHERE uri = $1
+        `, userInfo.Issuer)
+
+	var issuerID string
+
+	err := row.Scan(&issuerID)
+	switch err {
+	case nil:
+	case sql.ErrNoRows:
+		return nil, types.ErrorIssuerNotFound
+	default:
+		return nil, err
+	}
+
+	row = s.db.QueryRowContext(ctx, `
+        INSERT INTO user_info (name, email, sub, iss_id) VALUES (
+            $1, $2, $3, $4
+	) ON CONFLICT (sub, iss_id)
+        DO UPDATE SET
+            sub = excluded.sub, iss_id = excluded.iss_id
+        RETURNING id`,
```

Comment:
This insert is easier if I let the DB handle the ID. If the `(issuer, subject)` pair exists in the DB, what I get back is the ID for the existing user, otherwise i'll get a new row.

As for testing, I just create a user in the test setup process and reference that user if I need to test based on ID.